File: mobile/test/util/board_union_find_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:frontend/util/board_union_find.dart';

void main() {
  late BoardUnionFindind uf;

  setUp(() {
    uf = BoardUnionFindind(3);
  });

  test("simple", () {
    expect(uf.isSame(Pair(0, 0), Pair(1, 0)), equals(false));
    expect(uf.unite(Pair(0, 0), Pair(1, 0)), equals(true));
    expect(uf.unite(Pair(1, 0), Pair(0, 0)), equals(false));
    expect(uf.isSame(Pair(0, 0), Pair(1, 0)), equals(true));
    expect(uf.value(Pair(0, 0)), equals(3));
  });

  test("dango", () {
    expect(uf.isSame(Pair(0, 0), Pair(1, 1)), equals(false));
    expect(uf.unite(Pair(0, 0), Pair(1, 0)), equals(true));
    expect(uf.unite(Pair(1, 0), Pair(0, 0)), equals(false));
    expect(uf.unite(Pair(1, 0), Pair(1, 1)), equals(true));
    expect(uf.unite(Pair(1, 1), Pair(1, 0)), equals(false));
    expect(uf.unite(Pair(1, 1), Pair(0, 1)), equals(true));
    expect(uf.unite(Pair(0, 1), Pair(1, 1)), equals(false));
    expect(uf.unite(Pair(0, 1), Pair(0, 0)), equals(false));
    expect(uf.unite(Pair(0, 0), Pair(0, 1)), equals(false));
    expect(uf.isSame(Pair(0, 0), Pair(1, 1)), equals(true));
    expect(uf.value(Pair(0, 0)), equals(4));
  });
}

File: web/src/utils/boardUnionFind.test.ts
import { expect, test } from "vitest"
import { BoardUnionFind } from "./boardUnionFind"

test("simple", () => {
  const uf = new BoardUnionFind(3)
  expect(uf.isSame({ x: 0, y: 0 }, { x: 1, y: 0 })).toBe(false)
  expect(uf.unite({ x: 0, y: 0 }, { x: 1, y: 0 })).toBe(true)
  expect(uf.unite({ x: 1, y: 0 }, { x: 0, y: 0 })).toBe(false)
  expect(uf.isSame({ x: 0, y: 0 }, { x: 1, y: 0 })).toBe(true)
  expect(uf.getValue({ x: 0, y: 0 })).toBe(3)
})

test("dango", () => {
  const uf = new BoardUnionFind(3)
  expect(uf.isSame({ x: 0, y: 0 }, { x: 1, y: 1 })).toBe(false)
  expect(uf.unite({ x: 0, y: 0 }, { x: 1, y: 0 })).toBe(true)
  expect(uf.unite({ x: 1, y: 0 }, { x: 0, y: 0 })).toBe(false)
  expect(uf.unite({ x: 1, y: 0 }, { x: 1, y: 1 })).toBe(true)
  expect(uf.unite({ x: 1, y: 1 }, { x: 1, y: 0 })).toBe(false)
  expect(uf.unite({ x: 1, y: 1 }, { x: 0, y: 1 })).toBe(true)
  expect(uf.unite({ x: 0, y: 1 }, { x: 1, y: 1 })).toBe(false)
  expect(uf.unite({ x: 0, y: 1 }, { x: 0, y: 0 })).toBe(false)
  expect(uf.unite({ x: 0, y: 0 }, { x: 0, y: 1 })).toBe(false)
  expect(uf.isSame({ x: 0, y: 0 }, { x: 1, y: 1 })).toBe(true)
  expect(uf.getValue({ x: 0, y: 0 })).toBe(4)
})
